### DISCUSSION TITLE
How to perfrom collision check Whether there is collsion between a given volume box and FloatGrid ?

### DISCUSSION THREAD

**Discussion by jediofgever** [AcademySoftwareFoundation/openvdb#1120] [Q&A]
Hi I am wondering whats the most efficient way to perform collision checks within openvdb ? 
I look through whole documentation and API but didnt see anything related to it.

**Comment by Idclip** [comment]
It very much depends on the exact behaviour you're after. Is the `FloatGrid` in question a narrow band level set? Do you care about closest point information? Is activity intersection enough for your case?

There are a number of tools at your disposal to try. I'd recommend starting by taking a look at:

 - `openvdb/tools/FindActiveValues.h` https://www.openvdb.org/documentation/doxygen/FindActiveValues_8h.html
 - `openvdb/tools/RayIntersector.h` https://www.openvdb.org/documentation/doxygen/RayIntersector_8h.html

**Comment by jediofgever** [comment]
Yes the float grid is a level set.

I have a float grid representing of an outdoor envoirmemnt. The voxels sizes are about 0.1 meter. Then I have box with 6D pose, center postion and orientation. The box pose is randomly generated by some other parts of code, once a pose is generated then I want to look up whether the generated pose of box is in collision with envoirment voxels.

all of this voxels have a value that is assigned, 

if there is a collision between box and envoirment voxels, 
then I simply want return this value, otherwise some negative value.

**Comment by Idclip** [comment]
Right, so you also want to return the values of any voxels which overlap with a given domain. If it were a case of getting a yes/no answer to an intersection, then the tools in `openvdb/tools/FindActiveValues.h` might suffice. However it sounds like you need something a bit more custom. For example,. what happens if more than one voxel overlaps your domain? You're dealing with a narrow band level set, so is it possible for your domain to fall entirely in the inactive interior of the grid? You'd have to handle cases like this yourself.

A quick way to start might be to loop over all values in your `FloatGrid`, compute a corresponding world space bounds for each voxel and query if they fall inside the domain. A very naïve but trivial implementation would be to use something like `openvdb::tools::foreach` in `openvdb/tools/ValueTransformer.h`. 

**(I have not tested the below, this is just a starting point)**

```
FloatGrid g; // the grid
math::BBox<double> bounds; // 6D bounds

tools::foreach(g.beginValueOn(), [&](auto& value) {
    CoordBBox valuebounds;
    value->getBoundingBox(valuebounds);    
    auto ws = g.indexToWorld(valuebounds;) // get the word space bounds for this value
    if (bounds.hasOverlap(ws)) { /*boxes overlap*/ }
    else                                     { /*boxes do not overlap*/ }
});
```

To reiterate, there are plenty of optimisations that can be made here. You can leverage VDB's node hierarchy to quickly discard portions of the tree and more efficiently thread over individual branches. But hopefully this gives you some idea as to whether VDB is suitable for your use case.

**Comment by jediofgever** [comment]
THank you for your help. It is very useful. I Believe openvdb suits my needs here, thing is that I didnt want to reinvent the wheel and wanted to know whether there are existing tools out there to achieve collisions detection(finding intersections more precisely). Since the number of voxels I have are +1M, efficiency is extremely crucial. 

I was not aware of `FindActiveValues` class, seems like that uses node hierarchy to decide whether there is an overlap, which is the most efficient way ?  

For the clarity, I am only interested in active interiors of grid and if more than one volume intersects with domain, then simply take the interpolation  of intersecting voxels.
